electrum/channel_db.py

```diff
@@ -995,6 +1024,127 @@ def get_node_by_prefix(self, prefix):
+            if not potential_timestamps:
+                continue
+            chan_ann.timestamp = min(potential_timestamps)
+            timestamped_chan_anns.append(chan_ann)
+        return timestamped_chan_anns
+
+    def get_forwarding_gossip_batch(self) -> List[GossipForwardingMessage]:
+        with self.forwarding_lock:
+            fwd_gossip = self.fwd_channel_updates + self.fwd_node_announcements
+            channel_anns = self.fwd_channels.copy()
+            self.clear_forwarding_gossip()
+
+        fwd_chan_anns1, _ = self.filter_orphan_channel_anns(self.fwd_orphan_channels)
+        fwd_chan_anns2, self.fwd_orphan_channels = self.filter_orphan_channel_anns(channel_anns)
+        channel_anns = self.set_fwd_channel_anns_ts(fwd_chan_anns1 + fwd_chan_anns2)
+        return channel_anns + fwd_gossip
+
+    def get_gossip_in_timespan(self, timespan: GossipTimestampFilter) \
+        -> List[GossipForwardingMessage]:
+        """Return a list of gossip messages matching the requested timespan."""
+        forwarding_gossip = []
+        with self.lock:
+            chans = self._channels.copy()
+            policies = self._policies.copy()
+            nodes = self._nodes.copy()
+
+        for short_id, chan in chans.items():
+            # fetching the timestamp from the channel update (according to BOLT-07)
+            chan_up_n1 = policies.get((chan.node1_id, short_id))
+            chan_up_n2 = policies.get((chan.node2_id, short_id))
+            updates = []
+            for policy in [chan_up_n1, chan_up_n2]:
+                if policy and policy.raw and timespan.in_range(policy.timestamp):
+                    if policy.message_flags == 1:  # check that its not "dont_forward"
+                        updates.append(GossipForwardingMessage(
+                            msg=policy.raw,
+                            timestamp=policy.timestamp))
+            if not updates or chan.raw is None:
+                continue
+            chan_ann_ts = min(update.timestamp for update in updates)
+            channel_announcement = GossipForwardingMessage(msg=chan.raw, timestamp=chan_ann_ts)
+            forwarding_gossip.extend([channel_announcement] + updates)
+
+        for node_ann in nodes.values():
+            if timespan.in_range(node_ann.timestamp) and node_ann.raw:
+                forwarding_gossip.append(GossipForwardingMessage(
+                    msg=node_ann.raw,
+                    timestamp=node_ann.timestamp))
+        return forwarding_gossip
+
+    def get_channels_in_range(self, first_blocknum: int, number_of_blocks: int) \
+        -> List[ShortChannelID]:
+        with self.lock:
+            channels = self._channels.copy()
+        scids: List[ShortChannelID] = []
+        for scid in channels:
+            if first_blocknum <= scid.block_height < first_blocknum + number_of_blocks:
+                scids.append(scid)
+        scids.sort(key=lambda x: x.block_height)
```

Comment:
is the sort key needed? `ShortChannelID` just inherits `bytes`

Specifically, what about scids that are mined in the same block? Do we want them sorted into blockchain order? With this key they won't be. But without specifying any key, I think we would get that from the default sort order.

Reply:
added a test in https://github.com/spesmilo/electrum/commit/3de90952d1708026db584608b3e1c1267fea3a65 to confirm this